File: C++/Medium/Accepted/number-of-substrings-containing-all-three-characters_0.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        
        int n = s.size();
        int res = 0;
        int a = -1, b = -1, c = -1;
        for (int i = 0; i < n; ++i) {
            if (s[i] == 'a') {
                a = i;
            } else if (s[i] == 'b') {
                b = i;
            } else {
                c = i;
            }
            res += 1 + min(a, min(b, c));
        }
        return res;
    }
};

File: Rust/Medium/Accepted/number-of-substrings-containing-all-three-characters_0.rs
impl Solution {
    pub fn number_of_substrings(s: String) -> i32 {
        
        let mut cnt = [0; 3];
        let mut res = 0;
        let (mut l, mut r) = (0, 0);
        while r < s.len() {
            cnt[s.as_bytes()[r] as usize - 'a' as usize] += 1;
            while cnt[0] > 0 && cnt[1] > 0 && cnt[2] > 0 {
                cnt[s.as_bytes()[l] as usize - 'a' as usize] -= 1;
                l += 1;
            }
            res += l;
            r += 1;
        }
        res as i32
    }
}